### PR TITLE
Fix/legacy sst coexistence

### PR DESCRIPTION
**Pull request by plbossart**
Now that we have a mechanism upstream to deal with HDaudio legacy, let's use it.
Note that SOF still needs to enable the PCI class/subclass as listed in https://github.com/thesofproject/linux/issues/468

### CODE REVIEW

File: sound/pci/hda/hda_intel.c [hunk 1]
```diff
@@ -2132,9 +2135,26 @@ static int azx_probe(struct pci_dev *pci,
...
+			dev_info(&pci->dev, "No DSP detected, continuing HDaudio legacy probe\n");
+			break;
+		case SND_SKL_PCI_BIND_LEGACY:
+			dev_info(&pci->dev, "Module parameter forced binding with HDaudio legacy, bypassed detection logic\n");
+			break;
+		case SND_SKL_PCI_BIND_ASOC:
+			dev_info(&pci->dev, "Module parameter forced binding with SKL+ ASoC driver, aborting probe\n");
+			return -ENODEV;
+		default:
+			dev_err(&pci->dev, "invalid value for skl_pci_binding module parameter, ignored\n");
```
Comment (lyakh) [inline sound/pci/hda/hda_intel.c:2154]: Shouldn't ignoring an invalid value be equivalent to automatic detection, to also fall back to the pre-patch behaviour? Similar for the ASoC code below.

File: sound/pci/hda/Kconfig [hunk 2]
```diff
@@ -226,6 +226,68 @@ config SND_HDA_POWER_SAVE_DEFAULT
...
+# register for a PCI ID which is also handled by the HDAudio legacy
+# driver. When this option is selected and the DSP is detected based on
+# the PCI class/subclass/prog-if, the probe of the HDAudio legacy
+# aborts. This mechanism removes the need for distributions to use
+# blacklists. It can be bypassed with module parameters should the
+# Intel/Skylake or SOF drivers fail to handle a specific platform.
+
+config SND_HDA_INTEL_DSP_DETECTION_SKL
+	bool
+	help
```
Comment (lyakh) [inline sound/pci/hda/Kconfig:242]: wouldn't it be better to hide these Kconfig entries by removing their help entries, if they are anyway not for humans to toggle?

Comment (plbossart) [reply sound/pci/hda/Kconfig:242]: There are different opinions on this. I used comments in the past but was asked to add a help message.

File: sound/pci/hda/hda_intel.c (hunk 1 above)
Comment (plbossart) [reply sound/pci/hda/hda_intel.c:2154]: I don't get your point. In both cases an invalid module parameter is ignored and the previous behavior is used.

File: sound/pci/hda/Kconfig (hunk 2 above)
Comment (plbossart) [reply sound/pci/hda/Kconfig:242]: I also don't like the fact that commit messages are out of scope in github. Thanks for the typo report.

**Comment by plbossart** [comment]
Closing this PR for now. The autodetection isn't straighforward and until we support SKL and devices using the Intel private key this is premature.

**Pull request closed without merging**